File: api/v1/schemas/portfolio.py
# -*- coding: utf-8 -*-
"""
===================================
投资组合 API Schemas
===================================

职责：
1. 定义投资组合相关的 Pydantic 请求/响应模型
"""

from datetime import datetime, date
from decimal import Decimal
from typing import Optional, List, Dict, Any
from pydantic import BaseModel, Field, field_validator


class HoldingCreate(BaseModel):
    """持仓创建请求"""
    code: str = Field(..., description="股票代码")
    entry_price: float = Field(..., gt=0, description="建仓价格")
    weight: float = Field(..., gt=0, le=100, description="仓位占比")


class HoldingUpdate(BaseModel):
    """持仓更新请求"""
    entry_price: Optional[float] = Field(None, gt=0, description="建仓价格")
    weight: Optional[float] = Field(None, gt=0, le=100, description="仓位占比")


class HoldingResponse(BaseModel):
    """持仓响应"""
    id: int
    portfolio_id: Optional[int] = None
    code: str
    name: Optional[str] = None
    entry_price: float
    current_price: Optional[float] = None
    weight: float
    shares: Optional[float] = None
    pnl_pct: Optional[float] = None
    is_closed: bool = False
    added_at: Optional[str] = None
    updated_at: Optional[str] = None


class PortfolioCreate(BaseModel):
    """创建组合请求"""
    name: str = Field(..., min_length=1, max_length=100, description="组合名称")
    description: Optional[str] = Field(None, max_length=500, description="组合描述")
    initial_capital: Optional[float] = Field(None, gt=0, description="初始资金")
    currency: str = Field("CNY", description="货币代码")
    holdings: Optional[List[HoldingCreate]] = Field(None, description="初始持仓列表")


class PortfolioUpdate(BaseModel):
    """更新组合请求"""
    name: Optional[str] = Field(None, min_length=1, max_length=100, description="组合名称")
    description: Optional[str] = Field(None, max_length=500, description="组合描述")
    initial_capital: Optional[float] = Field(None, gt=0, description="初始资金")


class PortfolioSummary(BaseModel):
    """组合摘要"""
    total_value: float
    total_pnl_pct: float
    total_pnl_amount: float
    holdings_count: int


class PortfolioListItem(BaseModel):
    """组合列表项"""
    id: int
    name: str
    description: Optional[str] = None
    holdings_count: int
    total_value: float
    total_pnl_pct: float
    total_pnl_amount: float
    is_deleted: bool
    created_at: Optional[str] = None
    updated_at: Optional[str] = None


class PortfolioListResponse(BaseModel):
    """组合列表响应"""
    items: List[PortfolioListItem]
    total: int
    page: int
    limit: int


class PortfolioResponse(BaseModel):
    """组合响应"""
    id: int
    name: str
    description: Optional[str] = None
    initial_capital: Optional[float] = None
    currency: str
    is_deleted: bool
    created_at: Optional[str] = None
    updated_at: Optional[str] = None
    holdings: List[HoldingResponse] = []
    summary: Optional[PortfolioSummary] = None


class PortfolioDeleteResponse(BaseModel):
    """删除组合响应"""
    id: int
    name: str
    is_deleted: bool
    deleted_at: Optional[str] = None
    history_preserved: bool


class ClosePositionRequest(BaseModel):
    """平仓请求"""
    price_type: str = Field(..., description="价格类型: current/specified")
    specified_price: Optional[float] = Field(None, gt=0, description="指定价格")
    quantity_type: str = Field(..., description="数量类型: all/partial")
    quantity: Optional[float] = Field(None, gt=0, description="平仓数量")
    ratio: Optional[float] = Field(None, ge=0, le=100, description="平仓比例")


class ClosePositionPnl(BaseModel):
    """平仓盈亏信息"""
    amount: float
    percentage: float
    realized: bool


class ClosePositionRemaining(BaseModel):
    """剩余持仓信息"""
    shares: float
    weight: float


class ClosePositionResponse(BaseModel):
    """平仓响应"""
    id: int
    portfolio_id: int
    code: str
    name: Optional[str] = None
    entry_price: float
    close_price: float
    close_quantity: float
    close_type: str
    is_closed: bool
    closed_at: Optional[str] = None
    pnl: ClosePositionPnl
    remaining: ClosePositionRemaining
    weight_rebalance_required: bool = False


class BatchHoldingsRequest(BaseModel):
    """批量添加持仓请求"""
    holdings: List[HoldingCreate] = Field(..., description="持仓列表")
    rebalance_mode: str = Field("none", description="重平衡模式: none/equal/keep_existing")


class BatchHoldingsResponse(BaseModel):
    """批量添加持仓响应"""
    added_count: int
    skipped_count: int
    holdings: List[HoldingResponse]
    weight_summary: Dict[str, Any]


class RebalanceResponse(BaseModel):
    """平均分配仓位响应"""
    portfolio_id: int
    holdings_count: int
    new_weight_each: float
    holdings: List[Dict[str, Any]]


class CalculateResponse(BaseModel):
    """计算响应"""
    portfolio_id: int
    calculated_at: str
    snapshot_date: str
    total_value: float
    total_pnl_pct: float
    holdings_updated: int
    status: str


class HoldingSnapshotResponse(BaseModel):
    """持仓快照响应"""
    id: int
    history_id: int
    code: str
    name: Optional[str] = None
    entry_price: float
    current_price: Optional[float] = None
    weight: float
    pnl_pct: Optional[float] = None
    weighted_pnl: Optional[float] = None
    created_at: Optional[str] = None


class PortfolioHistoryItem(BaseModel):
    """历史记录项"""
    id: int
    portfolio_id: int
    snapshot_date: Optional[str] = None
    total_value: Optional[float] = None
    unrealized_pnl_pct: Optional[float] = None
    unrealized_pnl_amount: Optional[float] = None
    realized_pnl_amount: float = 0
    cumulative_realized_pnl: float = 0
    total_pnl_pct: Optional[float] = None
    total_pnl_amount: Optional[float] = None
    active_holdings_count: int = 0
    closed_holdings_count: int = 0
    calculation_status: str
    error_message: Optional[str] = None
    calculated_at: Optional[str] = None
    holdings: List[HoldingSnapshotResponse] = []


class PortfolioHistoryResponse(BaseModel):
    """历史记录响应"""
    portfolio_id: int
    items: List[PortfolioHistoryItem]
    total: int
    page: int
    limit: int


class PerformanceDataPoint(BaseModel):
    """收益走势数据点"""
    date: Optional[str] = None
    total_value: Optional[float] = None
    unrealized_pnl_pct: Optional[float] = None
    unrealized_pnl_amount: Optional[float] = None
    realized_pnl_amount: float = 0
    cumulative_realized_pnl: float = 0
    total_pnl_pct: Optional[float] = None


class PerformanceStatistics(BaseModel):
    """收益统计"""
    total_return_pct: Optional[float] = None
    unrealized_return_pct: Optional[float] = None
    realized_return_pct: Optional[float] = None
    max_drawdown_pct: Optional[float] = None
    best_day: Optional[Dict[str, Any]] = None
    worst_day: Optional[Dict[str, Any]] = None


class PerformanceResponse(BaseModel):
    """收益走势响应"""
    portfolio_id: int
    view_mode: str
    date_range: Dict[str, str]
    data_points: List[PerformanceDataPoint]
    statistics: PerformanceStatistics


class ErrorDetail(BaseModel):
    """错误详情"""
    field: Optional[str] = None
    constraint: Optional[str] = None


class PortfolioErrorResponse(BaseModel):
    """错误响应"""
    error: Dict[str, Any]

File: apps/dsa-web/src/api/portfolio.ts
import apiClient from './index';
import { toCamelCase } from './utils';

export interface Holding {
  id: number;
  portfolioId: number;
  code: string;
  name: string;
  entryPrice: number;
  lastPrice: number;
  currentPrice: number;
  weight: number;
  shares?: number;
  pnlPct: number;
  weightedPnl: number;
  isClosed: boolean;
  addedAt?: string;
  updatedAt?: string;
  closePrice?: number;
  closeQuantity?: number;
  closeType?: string;
  closedAt?: string;
}

export interface PortfolioSummary {
  totalValue: number;
  totalPnlPct: number;
  totalPnlAmount: number;
  holdingsCount: number;
}

export interface Portfolio {
  id: number;
  name: string;
  description?: string;
  initialCapital?: number;
  currency: string;
  isDeleted: boolean;
  createdAt?: string;
  updatedAt?: string;
  holdings: Holding[];
  summary: PortfolioSummary;
}

export interface PortfolioListItem {
  id: number;
  name: string;
  description?: string;
  holdingsCount: number;
  totalValue: number;
  totalPnlPct: number;
  totalPnlAmount: number;
  isDeleted: boolean;
  createdAt?: string;
  updatedAt?: string;
}

export interface CreatePortfolioRequest {
  name: string;
  description?: string;
  initialCapital?: number;
  currency?: string;
  holdings?: {
    code: string;
    name?: string;
    entryPrice: number;
    lastPrice?: number;
    weight: number;
  }[];
}

export interface UpdatePortfolioRequest {
  name?: string;
  description?: string;
  initialCapital?: number;
}

export interface ClosePositionRequest {
  priceType: 'current' | 'specified';
  specifiedPrice?: number;
  quantityType: 'all' | 'partial';
  quantity?: number;
  ratio?: number;
}

export interface ClosePositionResponse {
  id: number;
  portfolioId: number;
  code: string;
  name: string;
  entryPrice: number;
  closePrice: number;
  closeQuantity: number;
  closeType: string;
  isClosed: boolean;
  closedAt?: string;
  pnl: {
    amount: number;
    percentage: number;
    realized: boolean;
  };
  remaining: {
    shares: number;
    weight: number;
  };
  weightRebalanceRequired: boolean;
}

export interface BatchHoldingsRequest {
  holdings: {
    code: string;
    entryPrice: number;
    lastPrice?: number;
    weight: number;
  }[];
  rebalanceMode?: 'none' | 'equal' | 'keep_existing';
}

export interface BatchHoldingsResponse {
  addedCount: number;
  skippedCount: number;
  holdings: Holding[];
  weightSummary: Record<string, unknown>;
}

export interface RebalanceResponse {
  portfolioId: number;
  holdingsCount: number;
  newWeightEach: number;
  holdings: {
    id: number;
    code: string;
    weight: number;
  }[];
}

export interface CalculateResponse {
  portfolioId: number;
  calculatedAt: string;
  snapshotDate: string;
  totalValue: number;
  totalPnlPct: number;
  holdingsUpdated: number;
  status: string;
}

export interface PortfolioHistoryItem {
  id: number;
  portfolioId: number;
  snapshotDate: string;
  totalValue: number;
  unrealizedPnlPct: number;
  unrealizedPnlAmount: number;
  realizedPnlAmount: number;
  cumulativeRealizedPnl: number;
  totalPnlPct: number;
  totalPnlAmount: number;
  activeHoldingsCount: number;
  closedHoldingsCount: number;
  calculationStatus: string;
  errorMessage?: string;
  calculatedAt: string;
  holdings: Holding[];
}

export interface PerformanceDataPoint {
  date: string;
  totalValue: number;
  unrealizedPnlPct: number;
  unrealizedPnlAmount: number;
  realizedPnlAmount: number;
  cumulativeRealizedPnl: number;
  totalPnlPct: number;
  holdings?: Record<string, { pnlPct: number; weightedPnl: number }>;
}

export interface PerformanceStatistics {
  totalReturnPct: number;
  unrealizedReturnPct: number;
  realizedReturnPct: number;
  maxDrawdownPct: number;
  bestDay: { date: string; pnlPct: number };
  worstDay: { date: string; pnlPct: number };
}

export interface PerformanceResponse {
  portfolioId: number;
  viewMode: string;
  dateRange: {
    start: string;
    end: string;
  };
  dataPoints: PerformanceDataPoint[];
  statistics: PerformanceStatistics;
}

export const portfolioApi = {
  create: async (data: CreatePortfolioRequest): Promise<Portfolio> => {
    const requestData: Record<string, unknown> = {
      name: data.name,
    };
    if (data.description) requestData.description = data.description;
    if (data.initialCapital) requestData.initial_capital = data.initialCapital;
    if (data.currency) requestData.currency = data.currency;
    if (data.holdings) {
      requestData.holdings = data.holdings.map(h => ({
        code: h.code,
        name: h.name,
        entry_price: h.entryPrice,
        weight: h.weight,
      }));
    }

    const response = await apiClient.post<Record<string, unknown>>(
      '/api/v1/portfolios',
      requestData,
    );
    return toCamelCase<Portfolio>(response.data);
  },

  list: async (params?: {
    includeDeleted?: boolean;
    page?: number;
    limit?: number;
  }): Promise<{ items: PortfolioListItem[]; total: number; page: number; limit: number }> => {
    const queryParams: Record<string, string | number> = {};
    if (params?.includeDeleted) queryParams.include_deleted = String(params.includeDeleted);
    if (params?.page) queryParams.page = params.page;
    if (params?.limit) queryParams.limit = params.limit;

    const response = await apiClient.get<Record<string, unknown>>(
      '/api/v1/portfolios',
      { params: queryParams },
    );

    const data = toCamelCase<{
      items: PortfolioListItem[];
      total: number;
      page: number;
      limit: number;
    }>(response.data);

    return {
      items: (data.items || []).map(item => toCamelCase<PortfolioListItem>(item)),
      total: data.total,
      page: data.page,
      limit: data.limit,
    };
  },

  get: async (id: number): Promise<Portfolio> => {
    const response = await apiClient.get<Record<string, unknown>>(
      `/api/v1/portfolios/${id}`,
    );
    return toCamelCase<Portfolio>(response.data);
  },

  update: async (id: number, data: UpdatePortfolioRequest): Promise<Portfolio> => {
    const requestData: Record<string, unknown> = {};
    if (data.name !== undefined) requestData.name = data.name;
    if (data.description !== undefined) requestData.description = data.description;
    if (data.initialCapital !== undefined) requestData.initial_capital = data.initialCapital;

    const response = await apiClient.put<Record<string, unknown>>(
      `/api/v1/portfolios/${id}`,
      requestData,
    );
    return toCamelCase<Portfolio>(response.data);
  },

  delete: async (id: number): Promise<{ id: number; isDeleted: boolean }> => {
    const response = await apiClient.delete<Record<string, unknown>>(
      `/api/v1/portfolios/${id}`,
    );
    return toCamelCase<{ id: number; isDeleted: boolean }>(response.data);
  },

  restore: async (id: number): Promise<Portfolio> => {
    const response = await apiClient.post<Record<string, unknown>>(
      `/api/v1/portfolios/${id}/restore`,
    );
    return toCamelCase<Portfolio>(response.data);
  },

  addHolding: async (
    portfolioId: number,
    data: { code: string; entryPrice: number; weight: number; lastPrice?: number },
  ): Promise<Holding> => {
    const requestData = {
      code: data.code,
      entry_price: data.entryPrice,
      last_price: data.lastPrice,
      weight: data.weight,
    };

    const response = await apiClient.post<Record<string, unknown>>(
      `/api/v1/portfolios/${portfolioId}/holdings`,
      requestData,
    );
    return toCamelCase<Holding>(response.data);
  },

  updateHolding: async (
    portfolioId: number,
    holdingId: number,
    data: { entryPrice?: number; weight?: number },
  ): Promise<Holding> => {
    const requestData: Record<string, unknown> = {};
    if (data.entryPrice !== undefined) requestData.entry_price = data.entryPrice;
    if (data.weight !== undefined) requestData.weight = data.weight;

    const response = await apiClient.put<Record<string, unknown>>(
      `/api/v1/portfolios/${portfolioId}/holdings/${holdingId}`,
      requestData,
    );
    return toCamelCase<Holding>(response.data);
  },

  closePosition: async (
    portfolioId: number,
    holdingId: number,
    data: ClosePositionRequest,
  ): Promise<ClosePositionResponse> => {
    const requestData: Record<string, unknown> = {
      price_type: data.priceType,
      quantity_type: data.quantityType,
    };
    if (data.specifiedPrice) requestData.specified_price = data.specifiedPrice;
    if (data.quantity) requestData.quantity = data.quantity;
    if (data.ratio) requestData.ratio = data.ratio;

    const response = await apiClient.post<Record<string, unknown>>(
      `/api/v1/portfolios/${portfolioId}/holdings/${holdingId}/closePosition`,
      requestData,
    );
    return toCamelCase<ClosePositionResponse>(response.data);
  },

  batchAddHoldings: async (
    portfolioId: number,
    data: BatchHoldingsRequest,
  ): Promise<BatchHoldingsResponse> => {
    const requestData: Record<string, unknown> = {
      holdings: data.holdings.map(h => ({
        code: h.code,
        entry_price: h.entryPrice,
        weight: h.weight,
      })),
    };
    if (data.rebalanceMode) requestData.rebalance_mode = data.rebalanceMode;

    const response = await apiClient.post<Record<string, unknown>>(
      `/api/v1/portfolios/${portfolioId}/holdings/batch`,
      requestData,
    );
    return toCamelCase<BatchHoldingsResponse>(response.data);
  },

  rebalance: async (portfolioId: number): Promise<RebalanceResponse> => {
    const response = await apiClient.post<Record<string, unknown>>(
      `/api/v1/portfolios/${portfolioId}/holdings/rebalance`,
    );
    return toCamelCase<RebalanceResponse>(response.data);
  },

  calculate: async (portfolioId: number): Promise<CalculateResponse> => {
    const response = await apiClient.post<Record<string, unknown>>(
      `/api/v1/portfolios/${portfolioId}/calculate`,
    );
    return toCamelCase<CalculateResponse>(response.data);
  },

  getHistory: async (
    portfolioId: number,
    params?: {
      startDate?: string;
      endDate?: string;
      page?: number;
      limit?: number;
    },
  ): Promise<{ portfolioId: number; items: PortfolioHistoryItem[]; total: number; page: number; limit: number }> => {
    const queryParams: Record<string, string | number> = {};
    if (params?.startDate) queryParams.start_date = params.startDate;
    if (params?.endDate) queryParams.end_date = params.endDate;
    if (params?.page) queryParams.page = params.page;
    if (params?.limit) queryParams.limit = params.limit;

    const response = await apiClient.get<Record<string, unknown>>(
      `/api/v1/portfolios/${portfolioId}/history`,
      { params: queryParams },
    );

    const data = response.data as {
      portfolio_id: number;
      items: Record<string, unknown>[];
      total: number;
      page: number;
      limit: number;
    };

    return {
      portfolioId: data.portfolio_id || portfolioId,
      items: (data.items || []).map(item => toCamelCase<PortfolioHistoryItem>(item)),
      total: data.total,
      page: data.page,
      limit: data.limit,
    };
  },

  getPerformance: async (
    portfolioId: number,
    params?: {
      startDate?: string;
      endDate?: string;
      viewMode?: string;
    },
  ): Promise<PerformanceResponse> => {
    const queryParams: Record<string, string> = {};
    if (params?.startDate) queryParams.start_date = params.startDate;
    if (params?.endDate) queryParams.end_date = params.endDate;
    if (params?.viewMode) queryParams.view_mode = params.viewMode;

    const response = await apiClient.get<Record<string, unknown>>(
      `/api/v1/portfolios/${portfolioId}/performance`,
      { params: queryParams },
    );
    return toCamelCase<PerformanceResponse>(response.data);
  },

  export: async (
    portfolioId: number,
    params?: {
      format?: 'csv' | 'json';
      scope?: 'summary' | 'detailed';
      startDate?: string;
      endDate?: string;
    },
  ): Promise<Blob> => {
    const queryParams: Record<string, string> = {};
    if (params?.format) queryParams.format = params.format;
    if (params?.scope) queryParams.scope = params.scope;
    if (params?.startDate) queryParams.start_date = params.startDate;
    if (params?.endDate) queryParams.end_date = params.endDate;

    const response = await apiClient.get(
      `/api/v1/portfolios/${portfolioId}/export`,
      {
        params: queryParams,
        responseType: 'blob' as const,
      },
    );
    return response.data;
  },
};
